### PR TITLE
Public output for ZkProgram

### PR DESCRIPTION
**Pull request by mitschabaude**
Adapt the Pickles bindings to support public ouput in addition to public input.

Basically, all the changes here are natural consequences from a single line change in `snarky_js_bindings_lib.ml`:
```diff
 (* call into Pickles *)
  let tag, _cache, p, provers =
    Pickles.compile_promise () ~choices
-     ~public_input:(Input (public_input_typ public_input_size))
+     ~public_input:
+       (Input_and_output
+         ( public_input_typ public_input_size
+         , public_input_typ public_output_size ) )
      ~auxiliary_typ:Typ.unit
      ~branches:(module Branches)
      ~max_proofs_verified:(module Max_proofs_verified)
      ~name ~constraint_constants
```

### CODE REVIEW

File: ocaml/lib/snarky_js_bindings_lib.ml [hunk 1]
```diff
@@ -2133,7 +2187,9 @@ module Choices = struct
               raise_errorf
                 "Rules array is sparse; the entry at index %i is missing" index )
         in
-        let (Rule rule) = Inductive_rule.create ~public_input_size js_rule in
+        let (Rule rule) =
+          Inductive_rule.create ~public_input_size ~public_output_size js_rule
```
Comment (Trivo25) [inline ocaml/lib/snarky_js_bindings_lib.ml:2191]: just a question: What is a rule? Is it a "branch" of our program? (like a single method within the larger contract?)

**Review by Trivo25 — APPROVED** [review]
nice! Just a few questions to gain a better understanding of whats happening

File: ocaml/lib/snarky_js_bindings_lib.ml [hunk 2]
```diff
@@ -2004,22 +2053,22 @@ module Choices = struct
...
-                                 , b3
-                                 , b4 )
-                                 Pickles.Tag.t ) ->
-                            match Type_equal.Id.same_witness tag.id self.id with
-                            | None ->
-                                Pickles.Types_map.public_input tag
-                            | Some T ->
-                                public_input_typ public_input_size )
-                            tag self
+                        let get_typ (type a1 a2 a3 a4 b3 b4)
```
[the remark this answers is not in the document]
Comment (mitschabaude) [reply ocaml/lib/snarky_js_bindings_lib.ml:2056]: yeah, `typ` is a special term -- what we cal `Provable<T>` in snarkyjs

File: ocaml/lib/snarky_js_bindings_lib.ml (hunk 1 above)
Comment (mitschabaude) [reply ocaml/lib/snarky_js_bindings_lib.ml:2191]: yes, "rule" = method, also called "branches" or "choices" in other places :D

File: ocaml/lib/snarky_js_bindings_lib.ml [hunk 3]
```diff
@@ -2234,47 +2293,38 @@ let pickles_digest (choices : pickles_rule_js Js.js_array Js.t)
...
+        (Input_and_output
+           ( public_input_typ public_input_size
+           , public_input_typ public_output_size ) )
       ~auxiliary_typ:Typ.unit
       ~branches:(module Branches)
       ~max_proofs_verified:(module Max_proofs_verified)
       ~name ~constraint_constants
   in
+
+  (* translate returned prover and verify functions to JS *)
```
[the remark this answers is not in the document]
Comment (mitschabaude) [reply ocaml/lib/snarky_js_bindings_lib.ml:2317]: Fully agree!

**Pull request merged**